easybuild/easyblocks/c/cmake.py

```diff
@@ -47,6 +49,39 @@ def extra_options():
         })
         return extra_vars
 
+    def patch_step(self):
+        """
+        Apply patch files, and do runtime patching (if needed).
+
+        Tweak UnixPaths.cmake if EasyBuild is configured with --sysroot.
+        """
+        super(EB_CMake, self).patch_step()
+
+        sysroot = build_option('sysroot')
+        if sysroot:
+            # prepend custom sysroot to all hardcoded paths like /lib and /usr
+            # see also patch applied by Gentoo:
+            # https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-util/cmake/files/cmake-3.14.0_rc3-prefix-dirs.patch
```

Comment:
There are many small differences versus this patch. Those may not matter though but we should carefully check.

Reply:
I checked and all the small differences look inconsequential. Here '-' is the Gentoo-patched cmake and '+' the easyblock patch:
```
 list(APPEND CMAKE_SYSTEM_PREFIX_PATH
   # Standard
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/local /cvmfs/soft.computecanada.ca/gentoo/2020/usr /cvmfs/soft.computecanada.ca/gentoo/2020/
+  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/local /cvmfs/soft.computecanada.ca/gentoo/2020/usr /cvmfs/soft.computecanada.ca/gentoo/2020
```
this is just a slash

```
 # Non "standard" but common install prefixes
 list(APPEND CMAKE_SYSTEM_PREFIX_PATH
-  /usr/X11R6
-  /usr/pkg
-  /opt
+  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/X11R6
+  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/pkg
+  /cvmfs/soft.computecanada.ca/gentoo/2020/opt
```
these don't exist in the prefix, and most likely `/opt/bin`, `/opt/include` aren't used either so this looks harmless.

```
-list(APPEND CMAKE_SYSTEM_LIBRARY_PATH
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/lib//gcc
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/lib/
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/lib64
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/libx32
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/lib32
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/lib
-  /cvmfs/soft.computecanada.ca/gentoo/2020/lib
-  )
-
-list(APPEND CMAKE_SYSTEM_PROGRAM_PATH
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/bin
-  /cvmfs/soft.computecanada.ca/gentoo/2020/bin
+list(APPEND CMAKE_SYSTEM_LIBRARY_PATH /cvmfs/soft.computecanada.ca/gentoo/2020/usr/lib64 /cvmfs/soft.computecanada.ca/gentoo/2020/usr/lib /cvmfs/soft.computecanada.ca/gentoo/2020/lib64 /cvmfs/soft.computecanada.ca/gentoo/2020/lib
+  # X11
+ 
   )
```
The GCC paths are Gentoo GCC so not relevant to EB GCC.
The X11 path is a historical artifact, it's no longer relevant.
Appending to CMAKE_SYSTEM_PROGRAM_PATH on the Gentoo side looks silly since this defaults to CMAKE_SYSTEM_PREFIX_PATH entries suffixed with /bin already.

```
 list(APPEND CMAKE_PLATFORM_IMPLICIT_LINK_DIRECTORIES
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/lib//gcc
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/lib/
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/lib64
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/libx32
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/lib32
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/lib
-  /cvmfs/soft.computecanada.ca/gentoo/2020/lib
+  /cvmfs/soft.computecanada.ca/gentoo/2020/lib /cvmfs/soft.computecanada.ca/gentoo/2020/lib32 /cvmfs/soft.computecanada.ca/gentoo/2020/lib64 /cvmfs/soft.computecanada.ca/gentoo/2020/usr/lib /cvmfs/soft.computecanada.ca/gentoo/2020/usr/lib32 /cvmfs/soft.computecanada.ca/gentoo/2020/usr/lib64
   )
```
same story as above for GCC


```
@@ -92,11 +77,11 @@
 # parsing the implicit directory information from compiler output.
 set(_CMAKE_C_IMPLICIT_INCLUDE_DIRECTORIES_INIT
   ${CMAKE_C_IMPLICIT_INCLUDE_DIRECTORIES}
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/include
+  "${_cmake_sysroot_compile}/usr/include"
   )
 set(_CMAKE_CXX_IMPLICIT_INCLUDE_DIRECTORIES_INIT
   ${CMAKE_CXX_IMPLICIT_INCLUDE_DIRECTORIES}
-  /cvmfs/soft.computecanada.ca/gentoo/2020/usr/include
+  "${_cmake_sysroot_compile}/usr/include"
   )
 set(_CMAKE_CUDA_IMPLICIT_INCLUDE_DIRECTORIES_INIT
   ${CMAKE_CUDA_IMPLICIT_INCLUDE_DIRECTORIES}
```
not sure why Gentoo patches this. The implicit include directories are ok when they are initialized from the compiler output at CMake-compile-time, so we should not worry about this bit.